### DISCUSSION TITLE
Run error

### DISCUSSION THREAD

**Discussion by leyan88** [MarcEscandell/ALPypeRL#10] [General]
I am an anylogic beginner, download your code on GitHub, open the model in anylogic, but the model error:
![1](https://github.com/user-attachments/assets/c2e7db88-cd8b-44c1-bea9-54fff52fa82b)

**Comment by leyan88** [comment]
If it is convenient, I would like you to add a video using the explanatory model. As a beginner, it is a little difficult for me to learn from the documentation.

**Reply by MarcEscandell** [reply]
Hi @leyan88! Thanks for reaching out. If you are an AnyLogic beginner, maybe going straight into reinforcement learning might be a stretch :-D But let me try to help!

Following the documentation should be the best way to go, so I would like to insist to try to revisit the examples again (see [alpyperl docs](https://alpyperl.readthedocs.io/en/latest/)).

However, if you are really interested in just getting it running, here are my two cents: there are mainly two modes, _train_ ([train python file](https://github.com/MarcEscandell/ALPypeRL/blob/main/alpyperl/examples/cartpole_v0/train.py)) and _evaluate_ ([evaluate python file](https://github.com/MarcEscandell/ALPypeRL/blob/main/alpyperl/examples/cartpole_v0/evaluation.py)). In both of these cases, you need to have 2 things running: the _python script_ and the _AnyLogic model_ (in this exact order -you will be notified from python side when it's okay to run the _AnyLogic_ model). The reason why you get the error you shared is, probably, because you are just running the AnyLogic model. 

To be able to run the _python script_ you will be required to install _alpyperl_ first. For that, just run `pip install alpyperl`. Then, proceed to run one of the python scripts. The files shared assume that you are running the _CartPole-v0_ example. Depending on what _python_ file you are running, you will need to setup the mode in the _AnyLogic_ model from the `alPypeRLConnector` instance in the properties section.

**Reply by leyan88** [reply]
Thank you very much for your prompt response! I appreciate you taking the time to get back to me so quickly. I apologize for my delayed reply;  and I wanted to make sure I had enough time to properly review your project before responding.

I'm excited to hear that you are open to collaborating on integrating your project with AnyLogic. I believe that combining your work with AnyLogic could lead to some innovative applications, especially in the realm of decision-making and simulation.

Could you please provide more details on whether your project can be adapted to train DQN or DPG agents?

**Reply by MarcEscandell** [reply]
Since `alpyperl` is integrated with [Open AI Gymnasium](https://gymnasium.farama.org/) framework, you can use any RL model that is compatible. For example, if you use `rllib` (the recommended package, as `alpyperl` offers evaluation/deployment functionalities) here's a list of all [available models](https://docs.ray.io/en/latest/rllib/rllib-algorithms.html?_gl=1*r0snwr*_up*MQ..*_ga*MTg0MjQ3NjkwOC4xNzI0NzIzMzI1*_ga_0LCWHW1N3S*MTcyNDcyMzMyNS4xLjEuMTcyNDcyMzMyNy4wLjAuMA..).

In case of DQN, you can use the following python script for _training_ (as an example):

```python
from alpyperl import AnyLogicEnv
from ray.rllib.algorithms.dqn import DQNConfig


# Set checkpoint directory.
checkpoint_dir = "./resources/trained_policies/cartpole_v0"

# Initialize policy.
policy = (
    DQNConfig()
    .env_runners(
        num_env_runners=2,
        num_envs_per_env_runner=2
    )
    .fault_tolerance(
        recreate_failed_env_runners=True,
        num_consecutive_env_runner_failures_tolerance=3
    )
    .environment(
        AnyLogicEnv, 
        env_config={
            'run_exported_model': True,
            'exported_model_loc': './resources/exported_models/cartpole_v0',
            'show_terminals': False,
            'verbose': False,
            'checkpoint_dir': checkpoint_dir,
            'env_params': {
                'cartMass': 1.0,
                'poleMass': 0.1,
                'poleLength': 0.5,
            }
        }
    )
    .build()
)

# Perform training.
for _ in range(100):
    result = policy.train()

# Save policy checkpoint.
policy.save(checkpoint_dir)
print(f"Checkpoint saved in directory '{checkpoint_dir}'")

# Close all enviornments.
# NOTE: This is required to be called for correct checkpoint saving by ALPypeRL.
policy.stop()
```

**Reply by leyan88** [reply]
Thank you for your answers! I also want to know if ALpypeRL supports multi-agent reinforcement learning?

**Reply by MarcEscandell** [reply]
Hei! ALPypeRL does not support multi-agent RL at the moment. I have considered it as the next feature to include, but I haven't got a timeline. I'll keep you updated if I do so :)

**Reply by leyan88** [reply]
Thank you very much! This is a feature I am looking forward to and much needed！
